tidy3d/plugins/smatrix/smatrix.py

```diff
@@ -0,0 +1,267 @@
+"""Tools for generating an S matrix automatically from tidy3d simulation and port definitions."""
+
+from typing import List, Tuple, Dict, Any
+import os
+
+import pydantic as pd
+import numpy as np
+
+from ... import Simulation, Box, ModeSpec, ModeMonitor, ModeSource, GaussianPulse, SimulationData
+from ...constants import HERTZ, C_0
+from ...components.types import Direction, Ax
+from ...components.viz import add_ax_if_none, equal_aspect
+
+# fwidth of gaussian pulse in units of central frequency
+FWIDTH_FRAC = 1.0 / 10
+DEFAULT_DATA_DIR = "data"
+
+
+class Port(Box):
+    """Specifies a port in the scattering matrix."""
+
+    direction: Direction = pd.Field(
+        ...,
+        title="Direction",
+        description="'+' or '-', defining which direction is considered 'input'.",
+    )
+    mode_spec: ModeSpec = pd.Field(
+        ModeSpec(),
+        title="Mode Specification",
+        description="Specifies how the mode solver will solve for the modes of the port.",
+    )
+    mode_indices: Tuple[pd.NonNegativeInt] = pd.Field(
+        None,
+        title="Mode Indices.",
+        description="Indices into modes returned to mode solver to use in the port. "
+        "If ``None``, all modes returned by the mode solver are used in the scattering matrix. "
+        "Otherwise, the scattering matrix will include elements for each supplied mode index.",
+    )
+    name: str = pd.Field(
+        ...,
+        title="Name",
+        description="Unique name for the port.",
+        min_length=1,
+    )
+
+    @pd.validator("mode_indices", always=True)
+    def evaluate_mode_indices(cls, val, values):
+        """Evaluates mode indices based on number of modes in mode spec."""
+        if val is None:
+            num_modes = values.get("mode_spec").num_modes
+            val = tuple(range(num_modes))
+        return val
+
+
+"""
+s_matrix[port_name_in][port_name_out] gives a numpy array of shape (m, n)
+relating the coupling amplitudes between the m and n mode orders in the two ports, respectively.
+"""
+SMatrixType = Dict[str, np.ndarray]
+
+
+class ComponentModeler(pd.BaseModel):
+    """Tool for modeling devices and computing scattering matrix elements."""
+
+    simulation: Simulation = pd.Field(
+        ...,
+        title="Simulation",
+        description="Simulation describing the device without any sources or monitors present.",
+    )
+    ports: List[Port] = pd.Field(
+        [],
+        title="Ports",
+        description="Collection of ports describing the scattering matrix elements. "
+        "For each port, one siulation will be run with a modal source.",
+    )
+    freq: float = pd.Field(
+        ...,
+        title="Frequency",
+        description="Frequency at which to evaluate the scattering matrix.",
+        units=HERTZ,
+    )
+    batch_path: str = pd.Field(
+        "smatrix_batch.json",
+        title="Path to Batch",
```

Comment:
Is this related to the constant defined above, `DEFAULT_DATA_DIR`? If so, should the default path be `DEFAULT_DATA_DIR + "/smatrix_batch.json"`?

Reply:
No, it's a bit different.  This is where the batch metadata is saved, but the data for the batch (the individual `simulation_data.hdf5 files` are saved at DEFAULT_DATA_DIR.  We eventually need a better way to handle batches I think..